ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -47,16 +53,78 @@ var (
 	// headless svc will be of the form: {kuberay-cluster-name}-headless-worker-svc
 	headlessServiceSuffix = "headless-worker-svc"
 
-	// map of pod slices to workers in the slice
-	sliceToWorkers map[slice][]worker
-
 	// Flag arguments.
-	BindAddr   string
-	CACert     string
-	ServerCert string
-	ServerKey  string
+	BindAddr       string
+	CACert         string
+	KubeConfigPath string
+	ServerCert     string
+	ServerKey      string
 )
 
+func NewTPUWebhookServer(podLister listersv1.PodLister) *TPUWebhookServer {
+	return &TPUWebhookServer{
+		podLister: podLister,
+	}
+}
+
+func (t *TPUWebhookServer) Mutate(w http.ResponseWriter, r *http.Request) {
+	admissionReview := &admissionv1.AdmissionReview{}
+	if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+		http.Error(w, "Error decoding request body", http.StatusBadRequest)
+		return
+	}
+
+	if admissionReview.Request.Kind.Kind == "Pod" {
+		klog.V(1).Info("Received review for Pod creation")
+		response, err := mutatePod(t.podLister, admissionReview)
+		if err != nil {
+			klog.Errorf("Failed to mutate pod: %s", err)
+			return
+		}
+		admissionReview.Response = response
+		responseBytes, err := json.Marshal(admissionReview)
+		if err != nil {
+			klog.Errorf("Failed to encode response: %s", err)
+			return
+		}
+		fmt.Fprint(w, string(responseBytes))
+	}
+}
+
+func (t *TPUWebhookServer) Validate(w http.ResponseWriter, r *http.Request) {
+	admissionReview := &admissionv1.AdmissionReview{}
+	if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+		http.Error(w, "Error decoding request body", http.StatusBadRequest)
+		return
+	}
+
+	if admissionReview.Request.Kind.Kind == "RayCluster" {
+		klog.V(0).Info("Received review for RayCluster")
+		response, err := validateRayCluster(admissionReview)
+		if err != nil {
+			klog.Errorf("Failed to validate ray cluster: %s", err)
+			return
+		}
+		admissionReview.Response = response
+		responseBytes, err := json.Marshal(admissionReview)
+		if err != nil {
+			klog.Errorf("Failed to encode response: %s", err)
+			return
+		}
+		fmt.Fprint(w, string(responseBytes))
+	}
+}
+
+// helper function used for logging and testing
+func printSliceToWorkerIds(sliceToWorkerIDs map[slice][]int) {
```

Comment:
Where is this function called?

Reply:
It's not called anywhere right now, I was using it for debugging by adding it in `mutatePod`, `getSliceToWorkerIDs` etc. but I removed those lines before pushing the changes. Maybe I should move it to the test file instead?